compiler/rustc_middle/src/hir/map/mod.rs

```diff
@@ -707,6 +707,15 @@ impl<'hir> Map<'hir> {
                     Node::Block(Block { expr: None, .. }) => return None,
                     // The current node is not the tail expression of its parent.
                     Node::Block(Block { expr: Some(e), .. }) if hir_id != e.hir_id => return None,
+                    // The next node is an Fn and the current node is not its body
+                    // which means the node we started the iteration at is not part
+                    // of the function's body, but we're concerned here only with
+                    // nodes enclosed in the body (fixes #114918)
+                    Node::Item(Item { kind: ItemKind::Fn(_, _, body_id), .. })
```

Comment:
Can you test this with associated functions (both in impls and traits) and maybe see if there's a way to form a test with closures

Reply:
Thanks for the suggestion @compiler-errors. The issue was reproducing for impl and trait functions as well.

I've now extended the fix to all `Item` kinds returned by `get_return_block()` and added tests for trait and impl functions. However, could not find a way to test for closures as they do not support generics :-(